src/kixi/heimdall/service.clj

```diff
@@ -105,23 +108,79 @@
       (success {:message "Invalidated successfully"}))
     (fail "Invalid or expired refresh token provided")))
 
-(defn create-group
+(defn create-group-event
+  [session communications {:keys [group user] :as input}]
+  (let [group-ok? (spec/valid? :kixi.heimdall.schema/group-params group)
+        user-ok? (spec/valid? :kixi.heimdall.schema/user user)]
+    (if (and user-ok? group-ok?)
+      (and (comms/send-event! communications :kixi.heimdall/group-created "1.0.0" (update input :user #(select-keys % [:id :username]))) true)
+      false)))
+
+(defn- create-group
   [session {:keys [group user]}]
-  (when user
-    (let [group-id (:group-id (group/create! session {:name (:group-name group)}))]
-      (member/add-user-to-group session (java.util.UUID/fromString (:id user)) group-id "owner")
-      {:kixi.comms.event/key :kixi.heimdall/group-created
-       :kixi.comms.event/version "1.0.0"
-       :kixi.comms.event/payload {:group-id group-id}})))
+  (let [user-id  (java.util.UUID/fromString (:id user))
+        group-id (:group-id (group/create! session {:name (:group-name group)
+                                                    :user-id user-id}))]
+    (member/add-user-to-group session user-id group-id)
+    {:kixi.comms.event/key :kixi.heimdall/group-created
+     :kixi.comms.event/version "1.0.0"
+     :kixi.comms.event/payload {:group-id group-id}}))
 
 (defn new-user
-  [session params]
+  [session communications params]
   (let [credentials (select-keys params [:username :password])
         [ok? res] (user/validate credentials)]
     (if ok?
       (if (user/find-by-username session {:username (:username credentials)})
         (fail "There is already a user with this username.")
         (do
           (user/add! session credentials)
+          (comms/send-event! communications :kixi.heimdall/user-created "1.0.0" (select-keys params [:username]))
           (success {:message "User successfully created!"})))
       (fail (str "Please match the required format: " res)))))
+
+(defn- add-member
+  [session {:keys [user-id group-id]}]
+  (let [user-id  (java.util.UUID/fromString user-id)
+        group-id (java.util.UUID/fromString group-id)]
+    (member/add-user-to-group session user-id group-id)
+    {:kixi.comms.event/key :kixi.heimdall/member-added
```

Comment:
It looks like `add-member` is called when `:kixi.heimdall/member-added` is received, so it should return `nil`, not the same event (otherwise a loop will occur).